File: ProgrammingFundamentalsExercises/DictionariesLambdaAndLinqExe/07.PopulationCounter/PopulationCounter.cs
using System;
using System.Collections.Generic;
using System.Linq;

class PopulationCounter
{
    static void Main(string[] args)
    {
        string command = Console.ReadLine();
        Dictionary<string, long> cities = new Dictionary<string, long>();

        Dictionary<string, Dictionary<string, long>> countries = new Dictionary<string, Dictionary<string, long>>();
        Dictionary<string, Dictionary<string, long>> countriesSorted = new Dictionary<string, Dictionary<string, long>>();

        Dictionary<string, long> countriesPopulation = new Dictionary<string, long>();
        Dictionary<string, long> countriesPopulationSorted = new Dictionary<string, long>();

        while (command != "report")
        {
            string[] inputSplit = command.Split('|');
            string city = inputSplit[0];
            string country = inputSplit[1];
            long population = long.Parse(inputSplit[2]);
            cities[city] = population;

            if (countries.ContainsKey(country))
            {
                countries[country] = countries[country].Concat(cities).ToDictionary(pair => pair.Key, pair => pair.Value);
            }
            else
            {
                countries[country] = cities;
            }

            cities = new Dictionary<string, long>();
            command = Console.ReadLine();
        }

        foreach (KeyValuePair<string, Dictionary<string, long>> pair in countries)
        {
            cities = pair.Value;
            cities = cities.OrderByDescending(value => value.Key).ToDictionary(citiesPair => citiesPair.Key, citiesPair => citiesPair.Value);
            countriesSorted[pair.Key] = cities;
            long population = 0;

            foreach (KeyValuePair<string, long> citiesPair in cities)
            {
                population += citiesPair.Value;
            }

            countriesPopulation[pair.Key] = population;
            population = 0;
        }

        countriesPopulationSorted = countriesPopulation.OrderByDescending(n => n.Value).ToDictionary(n => n.Key, n => n.Value);

        foreach (KeyValuePair<string, long> pair in countriesPopulationSorted)
        {
            string country = pair.Key;
            long totalPopulation = pair.Value;
            cities = countriesSorted[country];

            Console.WriteLine($"{country} (total population: {totalPopulation})");
            
            foreach(KeyValuePair<string, long> citiesPair in cities)
            {
                Console.WriteLine($"=>{citiesPair.Key}: {citiesPair.Value}");
            }
        }
    }
}

File: JavaAdvanced/SetsAndMapsExercises/src/PopulationCounter.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;
import java.util.stream.Collectors;

public class PopulationCounter {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));

        String command = reader.readLine();

        LinkedHashMap<String, LinkedHashMap<String, Long>> countryCities = new LinkedHashMap<>();

        while (!command.equals("report")) {
            String[] inputArgs = command.split("\\|");

            String city = inputArgs[0];
            String country = inputArgs[1];
            long population = Long.parseLong(inputArgs[2]);

            if (!countryCities.containsKey(country)) {
                countryCities.put(country, new LinkedHashMap<>());

                countryCities.get(country).put(city, population);
            } else {
                if (!countryCities.get(country).containsKey(city)) {
                    countryCities.get(country).put(city, population);
                } else {
                    countryCities.get(country).put(city, countryCities.get(country).get(city) + population);
                }
            }

            command = reader.readLine();
        }

        ArrayList<Map.Entry<String, LinkedHashMap<String, Long>>> countryCitiesOrdered =
                countryCities.entrySet().stream()
                        .sorted(Comparator.<Map.Entry<String, LinkedHashMap<String, Long>>>comparingLong
                                (a -> a.getValue().values().stream().mapToLong(b -> b).sum()).reversed())
                        .collect(Collectors.toCollection(ArrayList::new));

        for (Map.Entry<String, LinkedHashMap<String, Long>> pair : countryCitiesOrdered) {

            ArrayList<Map.Entry<String, Long>> cityPopulation = pair.getValue().entrySet().stream()
                    .sorted(Comparator.<Map.Entry<String, Long>>comparingLong(a -> a.getValue()).reversed())
                    .collect(Collectors.toCollection(ArrayList::new));

            long totalPopulation = cityPopulation.stream().mapToLong(a -> a.getValue()).sum();

            System.out.printf("%s (total population: %d)%n", pair.getKey(), totalPopulation);

            for (Map.Entry<String, Long> cityPopulationPair : cityPopulation) {
                System.out.printf("=>%s: %d%n", cityPopulationPair.getKey(), cityPopulationPair.getValue());
            }
        }
    }
}